xarray/backends/api.py

```diff
@@ -710,21 +723,32 @@ def open_dataarray(
 
 
 def open_mfdataset(
-    paths,
-    chunks=None,
-    concat_dim=None,
-    compat="no_conflicts",
-    preprocess=None,
-    engine=None,
-    data_vars="all",
+    paths: str | Iterable[str | os.PathLike],
+    chunks: T_CHUNKS = None,
+    concat_dim: str
+    | DataArray
+    | Index
+    | Sequence[str]
+    | Sequence[DataArray]
+    | Sequence[Index]
+    | None = None,
+    compat: Literal[
+        "identical", "equals", "broadcast_equals", "no_conflicts", "override"
+    ] = "no_conflicts",
+    preprocess: Callable[[Dataset], Dataset] | None = None,
+    engine: T_ENGINE = None,
+    data_vars: Literal["all", "minimal", "different"] | list[str] = "all",
     coords="different",
-    combine="by_coords",
-    parallel=False,
-    join="outer",
-    attrs_file=None,
-    combine_attrs="override",
+    combine: Literal["by_coords", "nested"] = "by_coords",
+    parallel: bool = False,
+    join: Literal["outer", "inner", "left", "right", "exact", "override"] = "outer",
+    attrs_file: str | os.PathLike | None = None,
+    combine_attrs: Literal[
+        "drop", "identical", "no_conflicts", "drop_conflicts", "override"
+    ]
+    | Callable[..., Any] = "override",
     **kwargs,
-):
+) -> Dataset:
```

Comment:
Gosh, this is quite the type signature. Maybe in the future we can reduce the `concat_dim` into some sort of `IntoDim` generic. Thanks for working through the whole thing!

Reply:
I have moved some types to `core.types` which makes it already much more readable (and in the process opened a rabit hole of adding it everywhere...)

But: the `concat_dim` arg is not clear to me what the type is actually allowed.
The `Index` is only there because the docstring mentions it, in the code it is not checked so this might be actually wrong.
Someone with a better understanding of this part should check this type!